include/evmc/mocked_host.hpp

```diff
@@ -0,0 +1,258 @@
+    void set_balance(uint64_t x) noexcept
+    {
+        balance = uint256be{};
+        for (std::size_t i = 0; i < sizeof(x); ++i)
+            balance.bytes[sizeof(balance) - 1 - i] = static_cast<uint8_t>(x >> (8 * i));
+    }
+};
+
+class MockedHost : public Host
+{
+public:
+    struct log_record
+    {
+        address creator;
+        bytes data;
+        std::vector<bytes32> topics;
+
+        bool operator==(const log_record& other) const noexcept
+        {
+            return creator == other.creator && data == other.data &&
+                   std::equal(topics.begin(), topics.end(), other.topics.begin(),
+                              other.topics.end());
+        }
+    };
+
+    struct selfdestuct_record
+    {
+        address selfdestructed;
+        address beneficiary;
+
+        bool operator==(const selfdestuct_record& other) const noexcept
+        {
+            return selfdestructed == other.selfdestructed && beneficiary == other.beneficiary;
+        }
+    };
+
+    std::unordered_map<address, MockedAccount> accounts;
+
+    evmc_tx_context tx_context = {};
+
+    bytes32 blockhash = {};
+
+    evmc_result call_result = {};
+
+    std::vector<int64_t> recorded_blockhashes;
+
+    static constexpr auto max_recorded_account_accesses = 200;
+    std::vector<address> recorded_account_accesses;
+
+    static constexpr auto max_recorded_calls = 100;
+    std::vector<evmc_message> recorded_calls;
+
+    std::vector<log_record> recorded_logs;
+    std::vector<selfdestuct_record> recorded_selfdestructs;
+
+protected:
+    std::vector<bytes> m_recorded_calls_inputs;
+
+    void record_account_access(const address& addr)
+    {
+        if (recorded_account_accesses.empty())
+            recorded_account_accesses.reserve(max_recorded_account_accesses);
+
+        if (recorded_account_accesses.size() < max_recorded_account_accesses)
+            recorded_account_accesses.emplace_back(addr);
+    }
+
+    bool account_exists(const address& addr) noexcept override
+    {
+        record_account_access(addr);
+        return accounts.count(addr);
+    }
+
+    bytes32 get_storage(const address& addr, const bytes32& key) noexcept override
+    {
+        record_account_access(addr);
+        const auto ait = accounts.find(addr);
+        if (ait == accounts.end())
+            return {};
+
+        if (const auto sit = ait->second.storage.find(key); sit != ait->second.storage.end())
+            return sit->second.value;
+        return {};
+    }
+
+    evmc_storage_status set_storage(const address& addr,
+                                    const bytes32& key,
+                                    const bytes32& value) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return EVMC_STORAGE_UNCHANGED;
+
+        auto& old = it->second.storage[key];
+
+        // Follow https://eips.ethereum.org/EIPS/eip-1283 specification.
```

Comment:
2200 ?

Reply:
No.

Reply:
Well 1283 is kind of irrelevant, only existed in Constantinople before Petersburg.

I know technically it is the same except the stipend, but still.